### PR TITLE
feat: add app share property

### PR DESCRIPTION
**Pull request by deansheather**
Add `share` property to workspace app resources of type string. Must be one of `"owner"`, `"template"`, `"authenticated"`, `"public"`.

### CODE REVIEW

File: docs/resources/app.md [hunk 1]
```diff
@@ -26,11 +26,12 @@ EOF
...
-  agent_id  = coder_agent.dev.id
-  name      = "VS Code"
-  icon      = data.coder_workspace.me.access_url + "/icons/vscode.svg"
-  url       = "http://localhost:13337"
-  subdomain = false
+  agent_id      = coder_agent.dev.id
+  name          = "VS Code"
+  icon          = data.coder_workspace.me.access_url + "/icons/vscode.svg"
+  url           = "http://localhost:13337"
+  sharing_level = "owner"
```
Comment (kylecarbs) [inline docs/resources/app.md:33]: What do you think about `access` as a name instead?

Comment (deansheather) [reply docs/resources/app.md:33]: so just `access = "owner"`?

Comment (deansheather) [reply docs/resources/app.md]: after talking in slack we decided on `share` instead

File: provider/app.go [hunk 2]
```diff
@@ -76,6 +77,39 @@ func appResource() *schema.Resource {
 				ForceNew: true,
 				Optional: true,
 			},
+			"share": {
+				Type: schema.TypeString,
+				Description: "Determines the sharing level of the app. " +
```
Comment (kylecarbs) [inline provider/app.go:82]: I'd get rid of this first sentence. It's pretty implicit.

File: provider/app.go [hunk 3]
```diff
@@ -76,6 +77,39 @@ func appResource() *schema.Resource {
...
+					`Valid values are "owner", "template", "authenticated" ` +
+					`and "public". Level "owner" disables sharing on the ` +
+					"app, so only the workspace owner can access it. Level " +
+					`"template" shares the app with all users that can read ` +
+					`the workspace's template. Level "authenticated" shares ` +
+					`the app with all authenticated users. Level "public" ` +
+					"shares it with any user, including unauthenticated " +
+					"users. Permitted application sharing levels can be " +
+					`controlled via a flag on "coder server". Defaults to ` +
+					`"owner" (sharing disabled).`,
```
Comment (kylecarbs) [inline provider/app.go:95]: We should wait to allow the blocking of sharing globally IMO.

**Review by kylecarbs — APPROVED** [review]

File: provider/app.go (hunk 3 above)
Comment (deansheather) [reply provider/app.go]: Talked to Kyle on slack and he agrees we should keep the flag but it should default to all levels being available if you're entitled to the feature.

**Pull request merged**